design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+
+
+
+#### Environment & Timing Loop
+
+The design separates the timing loop from the `Adapter`. The while loop is built into an `Environment`, of which there is just one per simulation. We would like to be able to make updates to the `Environment` without having ot restart the simulation or renew the adapters. One possible implementation of the `Environment` might look like this:
+
+```python
+
+class Environment(object):
+
+    def __init__(self, adapter):
+        self._adapter = adapter # adapter wraps device (see above)
+        self._count = 0
+        self._timer = 0.0
+        self._delta = 0.0
+        self._user_delta = None
+        
+    def _delta(self):
+        if self._user_delta:
+            return self._user_delta
+        else:
+            return (datetime.now() - start).total_seconds()
+            
+    def set_user_delta(self, delta):
+        self._user_delta = delta
+
+    def run(self):
+
+        while True:
+            start = datetime.now()
+
+            
+            self._adapter.process(delta) # stream adapters would call asyncore.loop(0.1, count=1) as part of this
+
+            
+            count += 1
+            timer += self._delta()
+            self._delta = self._delta()
+            # ....
+```
+
+#### Environment Updates
+
+The separation of the runnable parts of the simulation (now in the `Environment`) from the `Device` and the `Adapter` parts should make updates which affect only the `Environment` easier.
+
+Here's one possibility I've been considering:
+
+`EnvironmentAdapter` sees `Environment` 
+
+```python 
+    class Environment(object):
+
+    def __init__(self, adapter, environ_adapter):
+        self._environ_adapter = environ_adapter
+        self._environ_adapter.set_adaptee(self)
+        self._adapter = adapter
+
+    def run(self):
+
+        while True:
+            # Everything still driven from a central processing loop.
+            self._environment_adapter.process()
+            
+            self._adapter.process(delta)
+
+```
+
+```python
+   class StreamEnvironmentAdapter(object):
+
+    def set_adaptee(self, environment):
+        self._environment = environment
+        
+    
+    def process(self):
+        asyncore.loop(0.1, count=1)
+        # ...
+        self._environment.set_user_delta(user_value)
+        
+```
+
+After discussion with Michael Wedel, I am now strongly leaning towards providing a simple single protocol for the `Environment Control`. While `Adapter` flexibility and variation is vital in order to support different device types. the `Environment` is internal to plankton and we should publish and control a single mechanism for this. Michael Wedel has suggested JSON over zeromq.
```

Comment:
If the problem is building JSON then I think we might be better off implementing the ZMQ+JSON anyway and provide a simple client. This is exactly what we did for the live data prototype [here](https://github.com/DMSC-Instrument-Data/live-data-prototype/blob/master/control.py).
Implementing something that a telnet client can talk to requires us to specify a protocol with a reasonable syntax that doesn't limit what we might want to do with this in the future. User input needs to be validated in some fashion as well. Because in the end we also expose the same API this way, it's just more complicated on the server side.